# 1. Solución Sistema de Ecuaciones Lineales

El objetivo de este notebook es resolver un problema de sistema de ecuaciones lineales. Evidenciar el impacto que puede causar en los resultados un pequeño cambio en los coeficiente del sistema y calcular la cota superior para el error relativo.

## Importando librerías necesarias

In [1]:
import numpy as np
from numpy import matrix, linalg

## 1.1. Hallar la solución para el siguiente sistema de Ecuaciones Lineales

In [2]:
#Definiendo la matriz con los coeficientes de las ecuaciones
A1=matrix([[2,4,5],[6,9,8],[4.1,5,3]])
A1

matrix([[2. , 4. , 5. ],
        [6. , 9. , 8. ],
        [4.1, 5. , 3. ]])

In [3]:
#Definiendo el vector con los resultados de cada ecuación
b=matrix([[220],[490],[274]])
b

matrix([[220],
        [490],
        [274]])

In [4]:
#Resolviendo el sistema de ecuaciones
solution_matrix1=linalg.solve(A1,b)
solution_matrix1

matrix([[40.],
        [10.],
        [20.]])

### Solución al sistema de ecuaciones

Los precios por hora para cada uno de los procesadores **A**, **B** y **C** son **40**, **10** y **20** respectivamente



## 1.2. Resolver el nuevo sistema de ecuaciones

In [5]:
A2=matrix([[2,4,5],[6,9,8],[4.2,5,3]])
A2

matrix([[2. , 4. , 5. ],
        [6. , 9. , 8. ],
        [4.2, 5. , 3. ]])

In [6]:
solution_matrix2=linalg.solve(A2,b)
solution_matrix2

matrix([[20.        ],
        [31.53846154],
        [10.76923077]])

### Solución al nuevo sistema de ecuaciones

Los precios por hora para cada uno de los procesadores **A**, **B** y **C** son **20**, **31.53846** y **10.76923** respectivamente.

### Analizando los cambios en los resultados.

Es evidente que un pequeño cambio en los coeficientes del sistema, causan un gran cambio en los precios por hora de cada procesador. Esto se debe a que la *eliminación gaussiana* es muy sensible a cambios en los coeficientes de la matriz, a causa de que las operaciones entre filas amplifican el impacto de los errores.

Por este motivo, hay que ser muy precisos en la definición de los sitemas de ecuaciones lineales en aplicaciones prácticas, para no obtener resultados sesgados.



## Calculando la Cota superior del error relativo

La cota superior del error relativo (`ε`) se puede definir como : **ε ≤ κ ⋅ δ**

Donde:
`k` es el número de condición de la matriz que se representa por ||A||₂ * ||A⁻¹||₂

`δ` es el error relativo en los coeficientes de la matriz que se representa por  ||A1 - A2||₂ / ||A1||₂



### Calculando la condición de la matriz

Podemos hacer el calculo paso a paso o utilizar la función de numpy `linalg.cond`


**Paso a paso**

In [7]:
# Realizando el calculo paso a paso

##Calculando la matriz inversa del sistema de ecuaciones origial
A1_I=A1.I
A1_I


matrix([[ 10.        , -10.        ,  10.        ],
        [-11.38461538,  11.15384615, -10.76923077],
        [  5.30769231,  -4.92307692,   4.61538462]])

In [8]:
## Calculando la norma 2 de A1
norm_A1 = np.linalg.norm(A1, ord=np.inf)  # en numpy la norma 2 es ord=infinito
norm_A1

23.0

In [9]:
## Calculando la norma 2 de A1_I
norm_A1_1=np.linalg.norm(A1_I, ord=np.inf)
norm_A1_1

33.307692307692335

In [10]:
## Calculando el número de condición de la matriz
k=norm_A1*norm_A1_1
k

766.0769230769237

**Usando la función de numpy**

In [11]:
cond_A1 = np.linalg.cond(A1, p=np.inf)  # Using norm 2
cond_A1

766.0769230769237

### Calculando el error relativo de coeficientes entre A1 y A2

In [12]:
delta = np.linalg.norm(A1 - A2, ord=np.inf) / np.linalg.norm(A1, ord=np.inf)
delta

0.004347826086956545

### Calculando el valor para `ε`

In [13]:
epsilon = cond_A1 * delta

print(f"Número de condición (κ) de la matriz A1: {cond_A1}")
print(f"Diferencia relativa (δ) entre A1 y A2: {delta}")
print(f"Cota superior de error relativo (ε): {epsilon}")

Número de condición (κ) de la matriz A1: 766.0769230769237
Diferencia relativa (δ) entre A1 y A2: 0.004347826086956545
Cota superior de error relativo (ε): 3.3307692307692514


## Conclusión

#### La cota superior del error relativo, nos indica qué tan sensible es el sistema de ecuaciones a cambios en los coeficientes. En este caso, significa que el sistema es altamente sensible dado que se estima que la solución puede ser hasta 333.07% superior o inferior al valor real. En los casos en los que no tenemos dos matrices para calcular la diferencia relativa, el número de condición de la matriz es un buen indicador para conocer si ésta está mal condicionada y, por lo tanto, nos permite evaluar qué tan precisa puede ser para estimar valores reales.